### PR TITLE
feature: Added PolandSpecProvider

### PR DESCRIPTION
**Pull request by DevAerial**
* Added PESEL generator for PolandSpecProvider
* Added NIP generator for PolandSpecProvider
* Added REGON generator for PolandSpecProvider
* Added tests for PESEL generator
* Added tests for NIP generator
* Added tests for REGON generator

Example usage:
```python
>>> from mimesis import Generic
>>> from mimesis.providers import Datetime
>>> from mimesis.enums import Gender
>>> from mimesis.builtins import PolandSpecProvider
>>> generic = Generic('pl-PL')
>>> generic.add_provider(PolandSpecProvider)
>>> # generate random PESEL
>>> generic.poland_provider.pesel()
'55112554471'
>>> # generate random NIP
>>> generic.poland_provider.nip()
'4627172367'
>>> # generate random REGON number
>>> generic.poland_provider.regon()
'213773823'
>>> birth_date = Datetime().datetime()
>>> # generate PESEL with specific birthday date
>>> generic.poland_provider.pesel(birth_date=birth_date)
'06121418234'
>>> # generate PESEL for male 
>>> generic.poland_provider.pesel(gender=Gender.MALE)
'07022852217'
```

### CODE REVIEW

File: mimesis/builtins/pl.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                    day = self.random.randint(1, 28)
+            else:
+                day = self.random.randint(1, 31)
+        else:
+            year = birth_date.date().year
+            month = birth_date.date().month
+        pesel_digits = list(map(int, '{:02d}'.format(year)))
+        if 1800 <= year <= 1899:
+            month += 80
+        elif 1900 <= year <= 1999:
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: Why just do we check it?

File: mimesis/builtins/pl.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        pesel_digits += list(map(int, '{:03d}'.format(series_number)))
+        if gender == Gender.MALE:
+            gender_digit = self.random.choice((1, 3, 5, 7, 9))
+        elif gender == Gender.FEMALE:
+            gender_digit = self.random.choice((0, 2, 4, 6, 8))
+        else:
+            gender_digit = self.random.choice(range(10))
+        pesel_digits += list(map(int, str(gender_digit)))
+        args = (9, 7, 3, 1, 9, 7, 3, 1, 9, 7)
+        sum_v = sum(map(lambda x: x[0] * x[1], zip(args, pesel_digits)))
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: List comprehensions are better to use here.

File: mimesis/builtins/pl.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        """The name of the provider."""
+
+        name = 'poland_provider'
+
+    def nip(self) -> str:
+        """Generate random valid 10-digit NIP.
+
+        :return: valid 10-digit NIP
+        """
+        nip_digits = list(map(int, str(self.random.randint(101, 998))))
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: I'm not sure that I understand why to do it like that. Can you, please, explain to me why not:

```python
[_ for _ in self.random.randint(101, 998)]
```

File: mimesis/builtins/pl.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        for _ in range(6):
+            nip_digits.append(self.random.randint(0, 9))
+        args = (6, 5, 7, 2, 3, 4, 5, 6, 7)
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: Bad name for the variable. What is actually this?

File: mimesis/builtins/pl.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    def pesel(self, birth_date: Optional[DateTime]=None,
+              gender: Optional[Gender]=None) -> str:
+        """Generate random 11-digit PESEL.
+
+        :param birth_date: initial birth date (optional)
+        :param gender: gender of person
+        :return: valid 11-digit PESEL
+        """
+        if not birth_date:
+            year = self.random.randint(1800, 2299) % 100
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: It's better to use `providers.Datetime` here.

File: mimesis/builtins/pl.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            month += 80
+        elif 1900 <= year <= 1999:
+            pass
+        elif 2000 <= year <= 2099:
+            month += 20
+        elif 2100 <= year <= 2199:
+            month += 40
+        elif 2200 <= year <= 2299:
+            month += 60
+        pesel_digits += list(map(int, '{:02d}'.format(month)))
```
Comment (lk-geimfari) [inline mimesis/builtins/pl.py]: List comprehension here too.

**Comment by lk-geimfari** [comment]
@DevAerial Thank for your contributions! Please, check out the comments.

**Comment by lk-geimfari** [comment]
Also, it would be awesome to add information about this provider to documentation.

**Comment by sobolevn** [comment]
I will do the second review round, @lk-geimfari thanks for mentioning me.

File: mimesis/builtins/pl.py (hunk 1 above)
Comment (DevAerial) [reply mimesis/builtins/pl.py]: This checks are needed due to the way polish PESEL identification number is created. Here a bit of explanation from (Wiki)[https://en.wikipedia.org/wiki/PESEL]:
>The PESEL system has been designed to cover five centuries. To distinguish people born in     different centuries, numbers are added to the MM field: 
>  * for birthdates between 1900 and 1999 - no change to MM field is made (see below)
>  * for other birthdates:
>      * 1800-1899 - month field number is increased by 80
>       * 2000-2099 - month + 20
>      * 2100-2199 - month + 40
>       * 2200-2299 - month + 60
>For example, a person born on December 24, 2002 would have a PESEL number starting with 023224 and person born on December 24, 1902 would have a PESEL number starting with 021224.

**Comment by lk-geimfari** [comment]
@sobolevn Okay, thank you!

File: mimesis/builtins/pl.py (hunk 4 above)
Comment (DevAerial) [reply mimesis/builtins/pl.py]: `args` are special coefficients which are multiplied with corresponding digits in `nip_digits` in order to generate checksum digit for NIP.

Comment (lk-geimfari) [reply mimesis/builtins/pl.py]: Can we use more informative name?

Comment (DevAerial) [reply mimesis/builtins/pl.py]: Is `nip_coeffecients` ok?

File: mimesis/builtins/pl.py [hunk 7]
```diff
@@ -90,9 +94,10 @@ def regon(self) -> str:
 
         :return: valid 9-digit REGON
         """
-        args = (8, 9, 2, 3, 4, 5, 6, 7)
+        regon_coefficients = (8, 9, 2, 3, 4, 5, 6, 7)
```
Comment (sobolevn) [inline mimesis/builtins/pl.py]: Typo here: `region_`

**Comment by lk-geimfari** [comment]
@DevAerial :ping_pong:

File: mimesis/builtins/pl.py (hunk 7 above)
Comment (DevAerial) [reply mimesis/builtins/pl.py]: It's not a typo: [REGON](https://pl.wikipedia.org/wiki/REGON).

**Comment by DevAerial** [comment]
I've added additional test for `pesel` function when using specific birthday date and fix bug discovered related to this new test.

File: mimesis/builtins/pl.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        :param gender: gender of person
+        :return: valid 11-digit PESEL
+        """
+        if not birth_date:
+            year = Datetime().year(1800, 2299) % 100
+            month = self.random.randint(1, 12)
+            if month in (4, 6, 9, 10):
+                day = self.random.randint(1, 30)
+            elif month == 2:
+                if isleap(year):
```
Comment (pywkm) [inline mimesis/builtins/pl.py]: That also means this check of `isleap` will fail for edge cases (`1800, 1900, 2100 and 2200` aren't leap, but `00` is).

Why not use `calendar.monthrange` method, and lines 51-57 replace with `day = self.random.randint(1, monthrange(year, month)[1])`?

(above logic is used also in `mimesis.providers.Datetime.date` so maybe better idea is to extract part drawing random date object to separate helper method in that provider, and use it here entirely?)
```
birth_date = birth_date or Datetime().random_date(1800, 2299)  # only proposition, maybe "date_object" would be better?
year = birth_date.date().year
month = birth_date.date().month
day = birth_date.date().day
...
```

File: mimesis/builtins/pl.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            else:
+                day = self.random.randint(1, 31)
+        else:
+            year = birth_date.date().year % 100
+            month = birth_date.date().month
+            day = birth_date.date().day
+        pesel_digits = list(map(int, '{:02d}'.format(year)))
+        if 1800 <= year <= 1899:
+            month += 80
+        elif 1900 <= year <= 1999:
```
Comment (pywkm) [inline mimesis/builtins/pl.py]: I agree with original review from @lk-geimfari "Why do we check it?". :) You misunderstood him then - he didn't ask for PESEL algorithm, but why do you check case for 1900-1999 if you `pass` it in next line. :) Why not just delete those two lines?

Comment (sobolevn) [reply mimesis/builtins/pl.py]: @pywkm awesome! Thank you for your review.

**Comment by pywkm** [comment]
I don't know how I can help here. Don't want to create the second pull request. If @DevAerial doesn't want to/can't continue work on this PR, maybe she/he can give me write access to her/his fork, so I can resolve conflicts and make changes on that branch?

**Pull request closed without merging**